[Model Averaging] Skip model averaging for the first K steps

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61207**
* #61206

Model averager now must be combined with post-localSGD DDP communication hook. It will skip model averaging for the first K steps, because post-localSGD communication hook will run global gradient averaging during this phase.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29523738](https://our.internmc.facebook.com/intern/diff/D29523738/)

Commit:
[Model Averaging] Skip model averaging for the first K steps

Model averager now must be combined with post-localSGD DDP communication hook. It will skip model averaging for the first K steps, because post-localSGD communication hook will run global gradient averaging during this phase.

Differential Revision: [D29523738](https://our.internmc.facebook.com/intern/diff/D29523738/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/algorithms/model_averaging/averagers.py:
```diff
@@ -6,7 +6,7 @@
-    Averages parameters periodically or during the warm-up stage.
+    Averages parameters periodically after the warm-up stage.
 
     This can be used for running `post-local SDG <https://arxiv.org/abs/1808.07217>`_,
```
nit: s/SDG/SGD

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1007,12 +1007,12 @@ def test_periodic_model_averager(self):
             tensor = torch.ones_like(param.data) * rank
             expected_avg_tensor = torch.ones_like(param.data) * sum(range(world_size)) / world_size
-            averager = averagers.PeriodicModelAverager(model, warmup_steps=10, period=4)
+            averager = averagers.PeriodicModelAverager(model, warmup_steps=12, period=4)
```
Shouldn't we have a test that combines PeriodicModelAverager with the local sgd hook?

SciPioneer replied:
Hmm, I prefer keeping the unit test self-contained here.

Let me add another integration test called `test_post_localSGD_periodic_averaging` that involves both the comm hook and the model averaging component in the next PR.

pritamdamania87 commented on torch/distributed/algorithms/model_averaging/averagers.py:
```diff
@@ -84,10 +88,10 @@ def __init__(
         at each iteration in the training loop.
         """
-        if self.step < self.warmup_steps or self.step % self.period == 0:
+        if self.step >= self.warmup_steps and self.step % self.period == 0:
```
If warmup steps is 101 and period is 4, is it correct that we don't average for steps 101 - 103 and then average on step 104? Shouldn't this check be `(self.step - self.warmup_steps) % self.period == 0` ?

SciPioneer replied:
Fixed and updated the test case to cover this case.

Commit:
Update on "[Model Averaging] Skip model averaging for the first K steps"


Model averager now must be combined with post-localSGD DDP communication hook. It will skip model averaging for the first K steps, because post-localSGD communication hook will run global gradient averaging during this phase.

Differential Revision: [D29523738](https://our.internmc.facebook.com/intern/diff/D29523738/)

[ghstack-poisoned]

Commit:
Update on "[Model Averaging] Skip model averaging for the first K steps"


Model averager now must be combined with post-localSGD DDP communication hook. It will skip model averaging for the first K steps, because post-localSGD communication hook will run global gradient averaging during this phase.

Differential Revision: [D29523738](https://our.internmc.facebook.com/intern/diff/D29523738/)

[ghstack-poisoned]

Commit:
Update on "[Model Averaging] Skip model averaging for the first K steps"


Model averager now must be combined with post-localSGD DDP communication hook. It will skip model averaging for the first K steps, because post-localSGD communication hook will run global gradient averaging during this phase.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29523738](https://our.internmc.facebook.com/intern/diff/D29523738/)

[ghstack-poisoned]
